eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs

```diff
@@ -0,0 +1,231 @@
+          [ injectFailure $
+              PPViewHashesDontMatch SNothing (tx ^. bodyTxL . scriptIntegrityHashTxBodyL)
+          ]
+
+  it "UnspendableUTxONoDatumHash" $ do
+    let lang = eraMaxLanguage @era
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+
+    txIn <- impAnn "Produce script at a txout with a missing datahash" $ do
+      let addr = Addr Testnet (ScriptHashObj scriptHash) StakeRefNull
+      let tx =
+            mkBasicTx mkBasicTxBody
+              & bodyTxL . outputsTxBodyL .~ [mkBasicTxOut addr (inject (Coin 10))]
+      let resetDataHash = dataHashTxOutL .~ SNothing
+      let resetTxOutDataHash =
+            bodyTxL . outputsTxBodyL
+              %~ ( \case
+                    h :<| r -> resetDataHash h :<| r
+                    _ -> error "Expected non-empty outputs"
+                 )
+
+      txInAt (0 :: Int)
+        <$> withPostFixup
+          (fixupResetAddrWits <$> resetTxOutDataHash)
+          (submitTx tx)
+
+    submitFailingTx
+      (mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn])
+      [injectFailure $ UnspendableUTxONoDatumHash [txIn]]
+
+  it "ExtraRedeemers" $ do
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    let prp = MintingPurpose (AsIx 2)
+    dt <- arbitrary
+    let tx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL .~ [txIn]
+            & witsTxL . rdmrsTxWitsL <>~ Redeemers (Map.singleton prp (dt, ExUnits 0 0))
+    let submit = submitFailingTx tx [injectFailure $ ExtraRedeemers [prp]]
+    if eraProtVerLow @era < natVersion @9
+      then -- PlutusPurpose serialization was fixed in Conway
+        withCborRoundTripFailures submit
+      else submit
+
+  it "Wrong network ID" $ do
+    submitFailingTx
+      (mkBasicTx mkBasicTxBody & bodyTxL . networkIdTxBodyL .~ SJust Mainnet)
+      [injectFailure $ WrongNetworkInTxBody Testnet Mainnet]
+
+  -- This already exists in Shelley.Imp.UtxowSpec
+  it "Missing required key witness" $ do
+    sender <- impLogToExpr freshKeyHash
+    receiver <- impLogToExpr freshAddr
+    let tx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . outputsTxBodyL .~ [mkBasicTxOut receiver (inject (Coin 10))]
+            & bodyTxL . reqSignerHashesTxBodyL .~ [sender]
+        removeWitness = witsTxL . addrTxWitsL %~ Set.filter ((sender /=) . witVKeyHash)
+    withPostFixup (pure . removeWitness) $
+      submitFailingTx tx [injectFailure $ MissingVKeyWitnessesUTXOW [sender]]
+
+  -- This already exists in Shelley.Imp.UtxowSpec
+  it "Missing 1-phase script witness" $ do
+    signer <- impLogToExpr freshKeyHash
+    let scriptHash = hashScript @era $ fromNativeScript $ RequireSignature signer
+    txIn <- produceScript scriptHash
+    submitFailingTx
+      (mkBasicTx $ mkBasicTxBody & inputsTxBodyL .~ [txIn])
+      [injectFailure $ MissingScriptWitnessesUTXOW [scriptHash]]
```

Comment:
I still need to check whether we need Alonzo-specific tests for these predicate faiures

Reply:
There's no code in Alonzo, Mary or Allegra that produces these failures, so I'll assume the test in Shelley is sufficient.